### PR TITLE
copy plugin: Support globs

### PR DESCRIPTION
**Pull request by sergiusens**
This also removes the call to `cp` and uses plain python APIs

Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

### CODE REVIEW

**Comment by sergiusens** [comment]
This is a partial extraction from https://github.com/ubuntu-core/snapcraft/pull/91

**Comment by sergiusens** [comment]
@chipaca care to review?

File: snapcraft/plugins/copy.py [hunk 1]
```diff
@@ -40,10 +42,24 @@ def schema(cls):
...
             if not os.path.lexists(src):
-                raise EnvironmentError('file "{}" missing'.format(src))
-            dst = os.path.join(self.installdir, dst)
-            os.makedirs(os.path.dirname(dst), exist_ok=True)
-            self.run(['cp', '--preserve=all', '-R', src, dst],
-                     cwd=os.getcwd())
+                # If it doesn't exist, check if it is a glob
+                filelst = glob.glob(src)
+                if not all(os.path.lexists(i) for i in filelst) or \
+                        len(filelst) is 0:
```
Comment (sergiusens) [inline snapcraft/plugins/copy.py]: `not filelst` might be better

File: snapcraft/plugins/copy.py [hunk 2]
```diff
@@ -40,10 +42,24 @@ def schema(cls):
...
+            dst = os.path.join(self.installdir, self.options.files[src])
+
             if not os.path.lexists(src):
-                raise EnvironmentError('file "{}" missing'.format(src))
-            dst = os.path.join(self.installdir, dst)
-            os.makedirs(os.path.dirname(dst), exist_ok=True)
-            self.run(['cp', '--preserve=all', '-R', src, dst],
-                     cwd=os.getcwd())
+                # If it doesn't exist, check if it is a glob
+                filelst = glob.glob(src)
```
Comment (sergiusens) [inline snapcraft/plugins/copy.py]: maybe glob everything in `self.options.files` first?

File: snapcraft/plugins/copy.py (hunk 1 above)
Comment (chipaca) [inline snapcraft/plugins/copy.py]: yep

File: snapcraft/plugins/copy.py [hunk 3]
```diff
@@ -40,10 +42,24 @@ def schema(cls):
...
+                if not all(os.path.lexists(i) for i in filelst) or \
+                        len(filelst) is 0:
+                    raise EnvironmentError('file "{}" missing'.format(src))
```
Comment (chipaca) [inline snapcraft/plugins/copy.py]: wouldn't that result in “file "foo/bar*" missing”, which might lead people to think you don't support globs?

File: snapcraft/plugins/copy.py (hunk 2 above)
Comment (chipaca) [inline snapcraft/plugins/copy.py]: before printing the error you mean?

File: snapcraft/plugins/copy.py (hunk 1 above)
Comment (chipaca) [inline snapcraft/plugins/copy.py]: in what case does glob return a nonexisting file?

File: snapcraft/plugins/copy.py [hunk 4]
```diff
@@ -39,11 +41,22 @@ def schema(cls):
         }
 
     def build(self):
-        for src in sorted(self.options.files):
-            dst = self.options.files[src]
-            if not os.path.lexists(src):
+        files = self.options.files
+        globs = {f: files[f] for f in files if not os.path.lexists(f)}
```
Comment (chipaca) [inline snapcraft/plugins/copy.py]: suggest filtering by glob.has_magic

**Comment by chipaca** [comment]
Looks good. To better improve the error you could filter files-that-don't-match with `glob.has_magic`, and change the error to "glob doesn't match" or equivalent-but-less-nerdy. :+1: either way.

**Comment by chipaca** [comment]
actually you could use has_magic just as you're outputting the error

**Pull request merged**